vispy/visuals/visual.py

```diff
@@ -101,3 +105,40 @@ def update(self):
         Emit an event to inform listeners that this Visual needs to be redrawn.
         """
         self.events.update()
+
+    def _get_hook(self, shader, name):
+        """Return a FunctionChain that Filters may use to modify the program.
+        
+        *shader* should be "frag" or "vert"
+        *name* should be "pre" or "post"
+        """
+        assert name in ('pre', 'post')
+        key = (shader, name)
+        if key in self._hooks:
+            return self._hooks[key]
+        
+        prog = getattr(self, '_program', None)
+        if prog is None:
+            raise NotImplementedError("%s shader does not implement hook '%s'"
+                                      % key)
+        hook = StatementList()
+        if shader == 'vert':
+            prog.vert[name] = hook
+        elif shader == 'frag':
+            prog.frag[name] = hook
+        self._hooks[key] = hook
+        return hook
+        
+    def attach(self, filter):
+        """Attach a Filter to this visual. 
+        
+        Each filter modifies the appearance or behavior of the visual.
+        """
+        filter._attach(self)
+        self._filters.add(filter)
```

Comment:
Circular reference. The Clipper filter stores the visual object, and the filters are also stored on the Visual. Either of these should probably be a weakref.